### PR TITLE
Implement LWG-3809 Is `subtract_with_carry_engine<uint16_t>` supposed to work?

### LINKED ISSUE #4162: LWG-3809 Is `subtract_with_carry_engine<uint16_t>` supposed to work?
LWG-3809 Is `subtract_with_carry_engine<uint16_t>` supposed to work?

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4162.

Also updates citations in the modified test file.

### CODE REVIEW

File: stl/inc/random
Comment (StephanTLavavej) [inline stl/inc/random:698]: Test failures:
```
random(698): error C2398: Element '1': conversion from 'unsigned __int64' to 'unsigned int' requires a narrowing conversion
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/random
Comment (jwakely) [reply stl/inc/random:701]: The seed value will be reduced _mod_ 2147483563 anyway, and the `subtract_with_carry_engine` uses the LCG engine to produce _n_ 32-bit values (where _n_ is the number of 32-bit pieces needed for a word of size _w_), so a 32-bit LCG is the right choice. Using a larger seed is possible, but has no benefit.

I suppose we could do `value % 2147483563` before the conversion to `UIntType`:

```c++
linear_congruential_engine<result_type,
                           40014u,0u,2147483563u> e(value == 0u ? default_seed : value % 2147483563u);
```

That would remove any dependence on the width of `uint_least32_t`.

Comment (MattStephanson) [reply stl/inc/random:701]: Thanks for your comment. I can write a new LWG issue explaining the situation and proposing that fix, unless there's a better way to handle it.

Comment (jwakely) [reply stl/inc/random:701]: That would be great, thanks.

File: stl/inc/random
Comment (StephanTLavavej) [inline stl/inc/random:902]: Above, I think we should change `subtract_with_carry() : _Mybase(default_seed) {}` to use `0u` for consistency.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I pushed very minor changes to the tr1 codepaths to make them consistent with the Standard codepaths, which I think will be far less confusing. I think this needs only one maintainer approval.

File: stl/inc/random
Comment (jovibor) [inline stl/inc/random:699]: Style:
Case inconsistency between `u` and `U` literals, or it doesn't matter?

Comment (frederick-vs-ja) [reply stl/inc/random:699]: It seems that the standard wording consistently uses lowercase (`u`) in [rand], while MSVC STL consistently uses uppercase (`U`)...

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for subtracting 1 from the number of remaining LWG issues! :joy_cat: :tada: :chart_with_downwards_trend:

File: stl/inc/random
Comment (MattStephanson) [reply stl/inc/random]: For reference: LWG-4014.

**Pull request merged**
